### PR TITLE
[CDF-22563] Update Resources mode

### PR DESCRIPTION
**Pull request by doctrino**
## Issue
```python
>>> from cognite.client import CogniteCLient
>>> from cognite.client.data_classes import FileMetadataWrite
>>> 
>>> _ = client.files.upload(FileMetadataWrite(external_id="my_file", name="the_file", metadata={"my_key": "my_value"})
>>> 
>>> updated = client.files.update(FileMetadataWrite(external_id="my_file", name="the_file", metadata=None)
>>> print(updated.metadata)
{'my_key': 'my_value'}
```

The user expect that the metadata should be removed. 

## Solution

Expose the `mode` parameter in all update methods.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,11 @@ Changes are grouped as follows
 - `Fixed` for any bug fixes.
 - `Security` in case of vulnerabilities.
 
+## [7.61.0] - 2024-09-17
+### Added
+- In all `.update` methods the user can now specify how non-update objects should be converted to update objects
+  using the new `mode` parameter.
```
Comment (haakonvt) [inline CHANGELOG.md:23]:
```suggestion
- All `update` methods now accept a new parameter `mode` that controls how non-update objects should be
  interpreted. For example, should we do a partial update or a full replacement.
```

File: cognite/client/_api/events.py [hunk 2]
```diff
@@ -555,18 +555,34 @@ def delete(
...
+        self,
+        item: Event | EventWrite | EventUpdate | Sequence[Event | EventWrite | EventUpdate],
+        mode: Literal["replace_ignore_null", "patch", "replace"] = "replace_ignore_null",
     ) -> Event | EventList:
         """`Update one or more events <https://developer.cognite.com/api#tag/Events/operation/updateEvents>`_
 
         Args:
             item (Event | EventWrite | EventUpdate | Sequence[Event | EventWrite | EventUpdate]): Event(s) to update
+            mode (Literal["replace_ignore_null", "patch", "replace"]): How to update data when a non-update
+                object is given (Events or -Write). If you use 'replace_ignore_null', only the fields
```
Comment (haakonvt) [inline cognite/client/_api/events.py:581]:
```suggestion
                object is given (Event or -Write). If you use 'replace_ignore_null', only the fields
```

File: docs/source/appendix.rst [hunk 3]
```diff
@@ -14,12 +12,128 @@ notes apply:
...
+    ...         name="New TS",
+    ...         metadata={"key": "value"}
+    ...     )
+    ... )
+    >>>
+    >>> updated = client.time_series.update(
+    ...     TimeSeriesWrite(
+    ...         external_id="new_ts",
+    ...         description="Updated description",
+    ...         metadata={"new": "entry"}
```
Comment (haakonvt) [inline docs/source/appendix.rst:49]: Let's do this to showcase that patch will replace keys that overlap
```suggestion
    ...         metadata={"key": "value", "another": "one"}
    ...     )
    ... )
    >>>
    >>> updated = client.time_series.update(
    ...     TimeSeriesWrite(
    ...         external_id="new_ts",
    ...         description="Updated description",
    ...         metadata={"key": "new value", "brand": "new"}
```

File: docs/source/appendix.rst [hunk 4]
```diff
@@ -14,12 +12,128 @@ notes apply:
...
+    ...     ),
+    ...     mode="patch"
+    ... )
+    >>> print(updated.dump())
+    {'externalId': 'new_ts', 'name': 'New TS', 'description': 'Updated description', 'metadata': {'key': 'value', 'new': 'entry'}}
```
Comment (haakonvt) [inline docs/source/appendix.rst:54]: I suggest we use `from pprint import pprint` to make this prettier 😄

File: docs/source/appendix.rst [hunk 5]
```diff
@@ -14,12 +12,128 @@ notes apply:
...
+    >>> updated = client.time_series.update(
+    ...     TimeSeriesWrite(
+    ...         external_id="new_ts",
+    ...         description="Updated description",
+    ...         metadata={"new": "entry"}
+    ...     ),
+    ...     mode="patch"
+    ... )
+    >>> print(updated.dump())
+    {'externalId': 'new_ts', 'description': 'Updated description', 'metadata': {'new': 'entry'}}
```
Comment (haakonvt) [inline docs/source/appendix.rst:80]: Lets add a comment explicitly saying "name was not in update but still got removed"

File: docs/source/appendix.rst [hunk 6]
```diff
@@ -14,12 +12,128 @@ notes apply:
...
+    ...     )
+    ... )
+    >>>
+    >>> updated = client.time_series.update(
+    ...     TimeSeriesWrite(
+    ...         external_id="new_ts",
+    ...         description="Updated description",
+    ...         metadata={"new": "entry"}
+    ...     ),
+    ...     mode="patch"
```
Comment (haakonvt) [inline docs/source/appendix.rst:103]:
```suggestion
    ...     mode="replace_ignore_null"
```

File: docs/source/appendix.rst [hunk 7]
```diff
@@ -14,12 +12,128 @@ notes apply:
...
```
Comment (haakonvt) [inline docs/source/appendix.rst:77]:
```suggestion
    ...     mode="replace"
```

File: docs/source/appendix.rst [hunk 8]
```diff
@@ -14,12 +12,128 @@ notes apply:
...
+    ...         metadata={"key": "value"}
+    ...     )
+    ... )
+    >>>
+    >>> updated = client.time_series.update(
+    ...     TimeSeriesWrite(
+    ...         external_id="new_ts",
+    ...         description="Updated description",
+    ...     ),
+    ...     mode="patch"
```
Comment (haakonvt) [inline docs/source/appendix.rst:131]:
```suggestion
    ...     mode="replace_ignore_null"
```

**Review by haakonvt — COMMENTED** [review]
Good changes! Really like the new docs 👌

File: docs/source/appendix.rst [hunk 9]
```diff
@@ -14,12 +12,173 @@ notes apply:
...
+    ...         description="Updated description",
+    ...         metadata={"key": "new value", "brand": "new"}
+    ...     ),
+    ...     mode="patch"
+    ... )
+    >>> pprint(updated.as_write().dump())
+    {'description': 'Updated description',
+     'externalId': 'new_ts',
+     'metadata': {'key': 'value', 'new': 'entry'},
+     'name': 'New TS'}
```
[the remark this answers is not in the document]
Comment (doctrino) [reply docs/source/appendix.rst:65]: That is intentional to make it clearer exactly what is happening.

File: docs/source/appendix.rst [hunk 10]
```diff
@@ -14,12 +12,173 @@ notes apply:
...
+    ...     ),
+    ...     mode="replace_ignore_null"
+    ... )
+    >>> pprint(updated.as_write().dump())
+    {'description': 'Updated description',
+     'externalId': 'new_ts',
+     'metadata': {'key': 'value'},
+     'name': 'New TS'}
+
+**Note** Since `metadata` was not specified in the update, it was not changed.
```
[the remark this answers is not in the document]
Comment (doctrino) [reply docs/source/appendix.rst:181]: This is the gotcha I got in Toolkit, so I think it is valuable.

**Pull request merged**